### PR TITLE
Issue #641 mask all packages

### LINKED ISSUE #641: [PLACEHOLDER] - for issue #641
This is to ensure that issue numbers in GitLab and GitHub are the same

### PR DESCRIPTION
**Pull request by luitjansl**
Fixes #641

# Description
the mask_all_package function now also masks the idomain array, and can be used with a mask
containing only a layer dimension, or a mask with only x and y  (or mesh2d_nFaces) dimensions.
Added tests.

# Checklist

- [X] Links to correct issue
- [X] Update changelog, if changes affect users
- [X] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [X] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: imod/mf6/dis.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ class StructuredDiscretization(Package):
         "idomain": (RegridderType.OVERLAP, "mode"),
     }
 
-    _skip_mask_arrays = ["bottom", "idomain"]
+    _skip_mask_arrays = [ "bottom"]
```
Comment (Manangka) [inline imod/mf6/dis.py:91]: Please remove the space between [ and "
Apples also to the disv class

File: imod/mf6/model.py [hunk 2]
```diff
@@ -495,17 +495,21 @@ def regrid_like(
...
         self,
-        domain: GridDataArray,
+        mask: GridDataArray,
     ):
         """
         This function applies a mask to all packages in a model. The mask must
         be presented as an idomain-like integer array that has 0 or negative
         values in filtered cells and positive values in active cells
         """
+        if any([d not in ["x", "y", "layer", "mesh2d_nFaces", "dx", "dy"] for d in mask.coords]):
```
Comment (Manangka) [inline imod/mf6/model.py:507]: what is d? Would 'coor', 'coordinate' or 'dimension' be a better name?

File: imod/mf6/model.py [hunk 3]
```diff
@@ -542,17 +546,18 @@ def _get_regridding_domain(
...
+                regridder = regridder_collection.get_regridder(
+                    regriddertype,
+                    function,
+                )
+                regridded_idomain = regridder.regrid(idomain)
+                if included_in_all is None:
+                    included_in_all = regridded_idomain
+                else:
+                    included_in_all = included_in_all.where(regridded_idomain.notnull())
+                    included_in_all = regridded_idomain.where(regridded_idomain <=0, other = included_in_all)
```
Comment (Manangka) [inline imod/mf6/model.py:560]: I haven't tested this snippet but could the following work?
```
regridders = [regridder_collection.get_regridder(regriddertype,function) for function in functionlist for regriddertype, functionlist in methods.items()]
included_in_all = [regridder.regrid(idomain) for regridder in regridders if regridder.regrid(idomain).notnull()]
```
or 
```
regridders = [regridder_collection.get_regridder(regriddertype,function) for function in functionlist for regriddertype, functionlist in methods.items()]
regridded_domains = [regridder.regrid(idomain) for regridder in regridders]
included_in_all = [regridded_domain for regridded_domain in regridded_domains if regridded_domain.notnull()]
```

File: imod/mf6/package.py [hunk 4]
```diff
@@ -533,17 +533,37 @@ def mask(self, domain: GridDataArray) -> Any:
         masked: Package
             The package with part masked.
         """
+
+        if any([d not in ["x", "y", "layer", "mesh2d_nFaces", "dx", "dy"] for d in domain.coords]):
```
Comment (Manangka) [inline imod/mf6/package.py:537]: Same comment as before: Can we use a better name then d?

File: imod/mf6/package.py [hunk 5]
```diff
@@ -533,17 +533,37 @@ def mask(self, domain: GridDataArray) -> Any:
         masked: Package
             The package with part masked.
         """
+
```
Comment (Manangka) [inline imod/mf6/package.py:536]: This method is becoming really big. With the nested if and for loops it becomes hard to read.
You can try to split it up
Below is some untested code which probably needs a little bit of finetuning to get it to work:

```
def mask(self, domain: GridDataArray) -> Any:
   ....
    def can_be_masked(var, da):
        return self.skip_masking_dataarray(var) or len(da.dims) == 0 or set(da.coords).issubset(["layer"])
    
    maskable_vars = [var for var in self.dataset.data_vars.keys() if can_be_masked(var, self.dataset[var])]
    non_maskable_vars = [var for var in self.dataset.data_vars.keys() if not can_be_masked(var, self.dataset[var])]
    
    for var in non_maskable_vars:
        masked[var] = self.dataset[var]
    
    spatial_maskable_vars = [var for var in maskable_vars if da.coords in (spatial_dims)]
    non_spatial_maskable_vars = [var for var in maskable_vars if da.coords not in (spatial_dims)]
    
    for var in spatial_maskable_vars:
        masked[var] = mask_spatial_var(var)
    
    for var in non_spatial_maskable_vars:
        masked[var] = mask_non_spatial_var(var)



def mask_spatial_var(var):
    if len(da.dims) < len(da.coords):
        if "layer" in da.coords and "layer" not in da.dims:
            array_domain = domain.sel(layer=da.coords["layer"])
        if "layer" not in da.coords and "layer" in array_domain.coords:
            array_domain = domain.isel(layer=0)

    if issubclass(da.dtype.type, numbers.Integral):
        if var == "idomain":
            return da.where(array_domain > 0, other=array_domain)
        else:
            return da.where(array_domain > 0, other=0)
    elif issubclass(da.dtype.type, numbers.Real):
        return da.where(array_domain > 0)
    else:
        raise TypeError(
            f"Expected dtype float or integer. Received instead: {da.dtype}"
        )


def mask_non_spatial_var(var):
    if da.values[()] is not None:
        if is_scalar(da.values[()]):
            return da.values[()]  # For scalars, such as options
        else:  # For example for arrays with only a layer dimension
            return da
    else:
        return None

```

**Review by Manangka — CHANGES_REQUESTED** [review]

File: docs/api/changelog.rst [hunk 6]
```diff
@@ -15,7 +15,9 @@ Added
   run code. Can be useful to plot data.
 - An API package was added. It can be added to both flow and transport models, and its 
 presence allows users to interact with libMF6.dll through its API.
-
+- The :func:`imod.mf6.model.mask_all_packages` now also masks the idomain array
+  of the model discretizaion, and can be used with a mask array without a layer
```
Comment (JoerivanEngelen) [inline docs/api/changelog.rst:19]: Typo:

```suggestion
  of the model discretization, and can be used with a mask array without a layer
```

File: imod/mf6/model.py (hunk 2 above)
Comment (JoerivanEngelen) [reply imod/mf6/model.py:507]: Xarray would use ``coord`` here,  I think that is the right naming.

File: imod/mf6/model.py [hunk 7]
```diff
@@ -495,17 +495,21 @@ def regrid_like(
...
+    def mask_all_packages(
         self,
-        domain: GridDataArray,
+        mask: GridDataArray,
     ):
         """
         This function applies a mask to all packages in a model. The mask must
         be presented as an idomain-like integer array that has 0 or negative
         values in filtered cells and positive values in active cells
         """
```
Comment (JoerivanEngelen) [inline imod/mf6/model.py:506]: Extend docstrings:

```suggestion
        """
        This function applies a mask to all packages in a model. The mask must
        be presented as an idomain-like integer array that has 0 or negative
        values in filtered cells and positive values in active cells
        
        Parameters
        ----------
        mask: xr.DataArray, xu.UgridDataArray of ints
            idomain-like integer array. 1 sets cells to active, 0 sets cells to inactive, 
            -1 sets cells to vertical passthrough
        """
```

File: imod/mf6/dis.py (hunk 1 above)
Comment (luitjansl) [reply imod/mf6/dis.py:91]: done

File: imod/mf6/package.py (hunk 5 above)
Comment (JoerivanEngelen) [reply imod/mf6/package.py:536]: I like this refactoring!

File: imod/tests/test_mf6/test_mf6_model_masking.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+    # Now set these cells to active in the mask
+    mask.values[17:22] = 1
+    unstructured_flow_model.mask_all_packages(mask)
+
+    # Masking should not activate inactive cells
+    assert np.all(unstructured_flow_model.domain.values[:, 17: 22] == inactivity_marker)
+
+
+def test_masking_masks_vertical_passthrough_cells(
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_model_masking.py:115]: Nice edge case!

File: imod/tests/test_mf6/test_mf6_model_masking.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+        print_flows=True,
+        validate=True,
+    )
+    structured_flow_model["well"] = well
+    
+    mask = deepcopy(structured_flow_model.domain.sel(layer=1))
+    mask = mask.drop_vars("layer")
+    mask.values[*mask_cells] = 0
+
+    cell_count = len(structured_flow_model.domain.x)*len(structured_flow_model.domain.y)*len(structured_flow_model.domain.layer)
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_model_masking.py:153]:
```suggestion
    cell_count = structured_flow_model.domain.size
```

File: imod/mf6/model.py [hunk 10]
```diff
@@ -495,17 +495,21 @@ def regrid_like(
...
 
-    def _mask_all_packages(
+    def mask_all_packages(
         self,
-        domain: GridDataArray,
+        mask: GridDataArray,
     ):
         """
         This function applies a mask to all packages in a model. The mask must
         be presented as an idomain-like integer array that has 0 or negative
```
Comment (JoerivanEngelen) [inline imod/mf6/model.py:504]: You mention "negative" values and "positive", have you tested if masking also works for -2 and 2? Otherwise just be more specific: mention -1 and 1.

**Review by JoerivanEngelen — CHANGES_REQUESTED** [review]
We made this function public API, this means we have to expose it in the corresponding API file. Please add this method under ``docs/api/mf6.rst`` -> ``GroundwaterFlowModel.mask_all_packages`` and ``GroundwaterTransportModel.mask_all_packages``

File: imod/mf6/package.py (hunk 4 above)
Comment (luitjansl) [reply imod/mf6/package.py]: done

File: imod/mf6/package.py (hunk 5 above)
Comment (luitjansl) [reply imod/mf6/package.py:538]: done

File: imod/mf6/model.py (hunk 3 above)
Comment (luitjansl) [reply imod/mf6/model.py:567]: not done. The first statement produces strange behavior:
to test why in the debugger I first tried a simpler expression:
`[str(functionlist) for regriddertype, functionlist in methods.items()]`
this produces : `["['mean', ..., 'mode']"]`
But then adding the second loop fails":
`[ str(function) for function in functionlist for regriddertype, functionlist in methods.items()]`
produces:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'functionlist' is not defined
```
Also,- the final statement in the suggestion makes included_in_all into a list which is not really what we want.

Comment (Manangka) [reply imod/mf6/model.py:567]: Could you try to get this to work?
It would greatly simply the code. Instead of having multiple if statements in a for loop in a for loop
It also makes it clearer what you are trying to do because each of the list comprehensions has a single goal (Instead of having the for loops do multiple things)

Comment (JoerivanEngelen) [reply imod/mf6/model.py:567]: > But then adding the second loop fails": `[ str(function) for function in functionlist for regriddertype, > functionlist in methods.items()]` produces:
> 
> ```
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> NameError: name 'functionlist' is not defined
> ```

For nested list comprehensions, you have to change the order into:
`[ str(function) for regriddertype, functionlist in methods.items() for function in functionlist]`

[See this explanation on how nested list comprehensions work](https://stackoverflow.com/a/20639246)

**Review by JoerivanEngelen — APPROVED** [review]
Approve in advance, but could you take a look if it is possible with my tip to follow Sunny's suggestions?

File: imod/mf6/model.py (hunk 3 above)
Comment (luitjansl) [reply imod/mf6/model.py]: done, now making a list of regridders. I kept the do-loop because include_in_all is not a list. Simplified the do-loop by initializing include_in_all upfront.

File: imod/mf6/package.py [hunk 11]
```diff
@@ -526,39 +526,56 @@ def mask(self, domain: GridDataArray) -> Any:
...
+        if is_scalar(da.values[()]):
+            return True
+        spatial_dims = ["x", "y", "mesh2d_nFaces", "layer"]
+        if not np.any( [coord in spatial_dims for coord in da.coords]):
+            return True
+        return False
+
+    def _mask_spatial_var(self, var: str, mask: GridDataArray)->GridDataArray:
+        da = self.dataset[var]
+        array_mask = mask
```
Comment (Manangka) [inline imod/mf6/package.py:560]: Line 560 to 565 could probably be a separate method. The name of that method then also makes clear what it is doing. Looking at the code it think it is correcting the mask for missing layers
```
def _mask_spatial_var(self, var: str, mask: GridDataArray) -> GridDataArray:
  array_mask = correct_for_missing_layers(da, mask)
  if issubclass(da.dtype.type, numbers.Integral):
      if var == "idomain":
          return da.where(array_mask > 0, other=array_mask)
      else:
           return da.where(array_mask > 0, other=0)
      elif issubclass(da.dtype.type, numbers.Real):
          return da.where(array_mask > 0)
      else:
          raise TypeError(
              f"Expected dtype float or integer. Received instead: {da.dtype}"
            ) 

def correct_for_missing_layers(da, mask):
  array_mask  = mask
   if len(da.dims) < len(da.coords):
            if "layer" in da.coords and "layer" not in da.dims:
                array_mask = mask.sel(layer=da.coords["layer"])
            if "layer" not in da.coords and "layer" in array_mask.coords:
                array_mask = mask.isel(layer=0)

  return array_mask  
```

**Review by Manangka — CHANGES_REQUESTED** [review]

File: imod/mf6/package.py [hunk 12]
```diff
@@ -526,39 +526,56 @@ def mask(self, domain: GridDataArray) -> Any:
...
-                        masked[var] = da.values[()]  # For scalars, such as options
-                    else:  # For example for arrays with only a layer dimension
-                        masked[var] = da
-                else:
-                    masked[var] = None
+                masked[var] = self._mask_spatial_var(var, mask)
 
         return type(self)(**masked)
 
+    def _is_copy_variable_on_masking(self, var: str, da: GridDataArray)->bool:
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply imod/mf6/package.py:548]: done ( is_eligible_for_copying)

File: imod/mf6/package.py (hunk 11 above)
Comment (luitjansl) [reply imod/mf6/package.py]: done

**Review by Manangka — APPROVED** [review]
Looks good
